source/task_scheduler.c

```diff
@@ -15,173 +15,116 @@
+    /* Execute all remaining tasks as CANCELED.
+     * Do this in a loop so that tasks scheduled by other tasks are executed */
+    while (aws_task_scheduler_has_tasks(scheduler, NULL)) {
+        s_run_all(scheduler, UINT64_MAX, AWS_TASK_STATUS_CANCELED);
     }
 
-    aws_priority_queue_clean_up(&scheduler->queue);
-    scheduler->alloc = NULL;
-    scheduler->clock = NULL;
+    aws_priority_queue_clean_up(&scheduler->timed_queue);
+    AWS_ZERO_STRUCT(scheduler);
 }
 
-static inline int s_get_next_task(
-    struct aws_task_scheduler *scheduler,
-    struct aws_task *task,
-    uint64_t run_before,
-    uint64_t *next_run_time) {
-    struct task_container *possible_task;
-    if (aws_priority_queue_top(&scheduler->queue, (void **)&possible_task)) {
-        if (AWS_ERROR_PRIORITY_QUEUE_EMPTY == aws_last_error()) {
-            return aws_raise_error(AWS_ERROR_TASK_SCHEDULER_NO_TASKS);
-        }
-        return AWS_OP_ERR;
-    }
-
-    if (next_run_time) {
-        *next_run_time = possible_task->timestamp;
-    }
-
-    if (possible_task->timestamp > run_before) {
-        return aws_raise_error(AWS_ERROR_TASK_SCHEDULER_NO_READY_TASKS);
-    }
-
-    struct task_container container;
-    if (aws_priority_queue_pop(&scheduler->queue, (void *)&container)) {
-        return AWS_OP_ERR;
-    }
-
-    *task = container.task;
-    if (next_run_time) {
-        if (aws_priority_queue_top(&scheduler->queue, (void **)&possible_task)) {
-            *next_run_time = 0;
-            if (AWS_ERROR_PRIORITY_QUEUE_EMPTY == aws_last_error()) {
-                return AWS_OP_SUCCESS;
-            }
-            return AWS_OP_ERR;
+bool aws_task_scheduler_has_tasks(const struct aws_task_scheduler *scheduler, uint64_t *next_task_time) {
+    uint64_t timestamp = UINT64_MAX;
+    bool has_tasks = false;
+
+    if (!aws_linked_list_empty(&scheduler->asap_list)) {
+        timestamp = 0;
+        has_tasks = true;
+    } else {
+        struct aws_task **task_ptrptr = NULL;
+        if (aws_priority_queue_top(&scheduler->timed_queue, (void **)&task_ptrptr) == AWS_OP_SUCCESS) {
+            timestamp = (*task_ptrptr)->timestamp;
+            has_tasks = true;
         }
-        *next_run_time = possible_task->timestamp;
     }
-    return AWS_OP_SUCCESS;
-}
-
-int aws_task_scheduler_next_task(struct aws_task_scheduler *scheduler, struct aws_task *task, uint64_t *next_run_time) {
 
-    uint64_t now;
-    if (scheduler->clock(&now)) {
-        return AWS_OP_ERR;
+    if (next_task_time) {
+        *next_task_time = timestamp;
     }
-
-    return s_get_next_task(scheduler, task, now, next_run_time);
+    return has_tasks;
 }
 
-int aws_task_scheduler_schedule_now(struct aws_task_scheduler *scheduler, struct aws_task *task) {
-
-    uint64_t now;
-    if (scheduler->clock(&now)) {
-        return AWS_OP_ERR;
-    }
+void aws_task_scheduler_schedule_now(struct aws_task_scheduler *scheduler, struct aws_task *task) {
+    assert(task);
+    assert(task->fn);
```

Comment:
`assert(scheduler)`. Assume everyone is dumb.

Reply:
:heart: assert :heart: